File: src/schemas/image.py
from __future__ import annotations
from typing import List, Optional
from pydantic import BaseModel, Field


class ImageAnalysis(BaseModel):
    """A single extracted fact or observation from an image."""
    category: str = Field(..., description="Category of the observation (e.g. object, text, scene, person, document)")
    description: str = Field(..., description="Concise description of what was observed")
    confidence: Optional[str] = Field(default=None, description="Confidence qualifier: high, medium, low")


class ImageResult(BaseModel):
    """Result of analysing an image through the ImageAgent."""
    description: str = Field(
        default="",
        description="Natural-language summary of the image contents",
    )
    observations: List[ImageAnalysis] = Field(
        default_factory=list,
        description="Structured list of observations extracted from the image",
    )

    @property
    def is_empty(self) -> bool:
        return not self.description.strip() and len(self.observations) == 0

File: xmem-go/internal/prompts/image.go
package prompts

import "strings"

var imageSystemPrompt = `You are an image analysis system for an AI assistant's memory layer.

## YOUR TASK
Analyse images provided by the user and extract structured observations that
can be stored as long-term memory. Focus on facts that would help the assistant
recall context in future conversations.

## WHAT TO EXTRACT

### 1. Objects & Scenes
- Key objects visible in the image
- Setting or environment (office, kitchen, outdoor, etc.)
- Notable colours, brands, or logos

### 2. Text & Documents
- Any readable text, labels, signs, or code snippets
- Document type if applicable (receipt, ticket, ID, screenshot, etc.)

### 3. People & Context
- Number of people, approximate activities
- Do NOT attempt to identify individuals by name
- Clothing, posture, or mood indicators if relevant

### 4. Technical Content
- Code, diagrams, architecture drawings, charts
- Data tables, graphs, dashboards

## OUTPUT FORMAT

Return your analysis in this exact format:

DESCRIPTION: <1-3 sentence natural-language summary of the image>

OBSERVATIONS:
- [category] observation text (confidence: high/medium/low)
- [category] observation text (confidence: high/medium/low)

### Categories
Use one of: object, text, scene, person, document, technical, other

### Rules
- Be factual and specific — avoid speculation
- Include 2-8 observations depending on image complexity
- If the image is empty, corrupted, or unreadable, return:
  DESCRIPTION: Unable to analyse image.
  OBSERVATIONS: (none)

## CRITICAL REMINDERS
- **PRIVACY FIRST** — Never attempt facial recognition or name individuals
- **BE SPECIFIC** — "Blue 2024 MacBook Pro on wooden desk" not "a laptop"
- **MEMORY-ORIENTED** — Focus on facts the assistant should remember later
- **SKIP TRIVIAL** — Ignore background noise, watermarks, or irrelevant artefacts
`

func BuildImageSystemPrompt() string {
	return imageSystemPrompt
}

func PackImageQuery(query, imageURL string) string {
	var parts []string

	if imageURL != "" {
		parts = append(parts, "<image>\n"+imageURL+"\n</image>")
	}

	if query != "" {
		parts = append(parts, "<user_query>\n"+query+"\n</user_query>")
	} else {
		parts = append(parts, "<user_query>\nAnalyse this image.\n</user_query>")
	}

	parts = append(parts, "Analyse the image above. Return a DESCRIPTION and OBSERVATIONS following the specified format.")

	return strings.Join(parts, "\n\n")
}
